src/server/pfs/s3/driver.go

```diff
@@ -213,3 +228,62 @@ func (d *WorkerDriver) bucketCapabilities(pc *client.APIClient, r *http.Request,
 func (d *WorkerDriver) canModifyBuckets() bool {
 	return false
 }
+
+// LocalDriver is the driver for the s3gateway instance running on pachd
+// workers
+type LocalDriver struct {
+	localFS
+	root string
+}
+
+// NewLocalDriver creates a new worker driver. `inputBuckets` is a list of
+// whitelisted buckets to be served from input repos. `outputBucket` is the
+// whitelisted bucket to be served from an output repo. If `nil`, no output
+// bucket will be available.
+func NewLocalDriver(root string) *LocalDriver {
+	return &LocalDriver{
+		root: root,
+	}
+}
+
+func (d *LocalDriver) listBuckets(pc *client.APIClient, r *http.Request, buckets *[]*s2.Bucket) error {
+	bucketPaths, err := filepath.Glob(filepath.Join(d.root, "*"))
+	if err != nil {
+		return err
+	}
+
+	for _, path := range bucketPaths {
+		*buckets = append(*buckets, &s2.Bucket{
+			Name: filepath.Base(path),
+		})
+	}
+	return nil
+}
+
+func (d *LocalDriver) bucket(pc *client.APIClient, r *http.Request, name string) (*Bucket, error) {
+	return &Bucket{
+		Name: name,
+		Path: filepath.Join(d.root, name),
+	}, nil
+}
+
+func (d *LocalDriver) bucketCapabilities(pc *client.APIClient, r *http.Request, bucket *Bucket) (bucketCapabilities, error) {
+	_, err := os.Stat(bucket.Path)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			err = s2.NoSuchBucketError(r)
+		} else {
+			err = s2.InternalError(r, err)
+		}
+		return bucketCapabilities{}, err
+	}
+	return bucketCapabilities{
+		readable:         true,
```

Comment:
Yes, I think that will be necessary unless it hardcodes different treatment of an `out` bucket. I wanted to avoid that for now to not have to worry about differentiating between local and pfs per-bucket (in particular, not writing `CopyObject` yet) until I have the code that actually uses this.